### PR TITLE
Make sv_dump() (and Devel::Peek) show hex instead of octal.

### PR DESCRIPTION
**Pull request by FGasper**
This improves the readability of SV dumps.

Thank you!

### CODE REVIEW

**Comment by demerphq** [comment]
I have some concerns about this, we used octal for byte data and hex for
unicode. Differentiating the two is highly helpful in debugging things like
the regex engine. I would prefer this patch is not merged until I can
understand the impact on debugging the regex engine. Fwiw I created at
least one of these functions.

Yves

**Comment by Grinnz** [comment]
On Fri, Jan 28, 2022 at 10:30 PM Yves Orton ***@***.***>
wrote:

> I have some concerns about this, we used octal for byte data and hex for
> unicode. Differentiating the two is highly helpful in debugging things like
> the regex engine. I would prefer this patch is not merged until I can
> understand the impact on debugging the regex engine. Fwiw I created at
> least one of these functions.


That seems of limited utility as byte and unicode data cannot be
differentiated between the codepoints of 128-255.

-Dan

**Comment by demerphq** [comment]
Sorry for top posting,  wanted to clarify that I am not necessarily against
this change. If it does not hinder debugging hard things then I would be
happy with the change, all I ask is that we defer until I can understand
it's impact.

Yves

**Comment by demerphq** [comment]
On Sat, 29 Jan 2022, 11:31 Dan Book, ***@***.***> wrote:

> On Fri, Jan 28, 2022 at 10:30 PM Yves Orton ***@***.***>
> wrote:
>
> > I have some concerns about this, we used octal for byte data and hex for
> > unicode. Differentiating the two is highly helpful in debugging things
> like
> > the regex engine. I would prefer this patch is not merged until I can
> > understand the impact on debugging the regex engine. Fwiw I created at
> > least one of these functions.
>
>
> That seems of limited utility as byte and unicode data cannot be
> differentiated between the codepoints of 128-255.
>

If non utf8 octets are encoded as octal then I can tell that the data is
not intended to be interpreted as utf8. If those octets are displayed in
hex then I know otherwise. It aids comprehension on debugging complex
matters of encoding.

Also I assume you meant codepoints 0-127, not 128-255. But those are
codepoints, not octets. If the octets are coming from a buffer known to be
utf8 then it provides a visual differentiation.

Anyway, all I am asking is to hold off on this until I can understand what
impact it has on scenarios for which the functions were written to serve.
It may be it has no impact or is an improvement in which case I have no
issue. It also could be that my memory fails me and I'm thinking of
something else. I think my request is quite reasonable given I am the
author of some of the code involved.

Yves

**Comment by Grinnz** [comment]
On Fri, Jan 28, 2022 at 10:59 PM Yves Orton ***@***.***>
wrote:

> On Sat, 29 Jan 2022, 11:31 Dan Book, ***@***.***> wrote:
>
> > On Fri, Jan 28, 2022 at 10:30 PM Yves Orton ***@***.***>
> > wrote:
> >
> > > I have some concerns about this, we used octal for byte data and hex
> for
> > > unicode. Differentiating the two is highly helpful in debugging things
> > like
> > > the regex engine. I would prefer this patch is not merged until I can
> > > understand the impact on debugging the regex engine. Fwiw I created at
> > > least one of these functions.
> >
> >
> > That seems of limited utility as byte and unicode data cannot be
> > differentiated between the codepoints of 128-255.
> >
>
> If non utf8 octets are encoded as octal then I can tell that the data is
> not intended to be interpreted as utf8. If those octets are displayed in
> hex then I know otherwise. It aids comprehension on debugging complex
> matters of encoding.
>
> Also I assume you meant codepoints 0-127, not 128-255. But those are
> codepoints, not octets. If the octets are coming from a buffer known to be
> utf8 then it provides a visual differentiation.


That is not correct. Codepoints 128-255 could represent either bytes or
characters. There is no differentiation stored and the representation does
not indicate the user's intent. To depend on it being one or the other
without knowing intent is the unicode bug.

-Dan

**Comment by demerphq** [comment]
On Sat, 29 Jan 2022, 12:04 Dan Book, ***@***.***> wrote:

> On Fri, Jan 28, 2022 at 10:59 PM Yves Orton ***@***.***>
> wrote:
>
> > On Sat, 29 Jan 2022, 11:31 Dan Book, ***@***.***> wrote:
> >
> > > On Fri, Jan 28, 2022 at 10:30 PM Yves Orton ***@***.***>
> > > wrote:
> > >
> > > > I have some concerns about this, we used octal for byte data and hex
> > for
> > > > unicode. Differentiating the two is highly helpful in debugging
> things
> > > like
> > > > the regex engine. I would prefer this patch is not merged until I can
> > > > understand the impact on debugging the regex engine. Fwiw I created
> at
> > > > least one of these functions.
> > >
> > >
> > > That seems of limited utility as byte and unicode data cannot be
> > > differentiated between the codepoints of 128-255.
> > >
> >
> > If non utf8 octets are encoded as octal then I can tell that the data is
> > not intended to be interpreted as utf8. If those octets are displayed in
> > hex then I know otherwise. It aids comprehension on debugging complex
> > matters of encoding.
> >
> > Also I assume you meant codepoints 0-127, not 128-255. But those are
> > codepoints, not octets. If the octets are coming from a buffer known to
> be
> > utf8 then it provides a visual differentiation.
>
>
> That is not correct. Codepoints 128-255 could represent either bytes or
> characters. There is no differentiation stored and the representation does
> not indicate the user's intent. To depend on it being one or the other
> without knowing intent is the unicode bug.
>

SThese functions are designed to debug the internal state of our
datstructures. They are are called from places where the internals very
much need to know if they are dealing with ut8 encoded data or not.

Codepoints 0-127 are indistinguishable at both octets and codepoint level.
Octets 128-255 may represent codepoints 128-255  or they may represent
parts of an utf8 sequence which represents a codepoint larger than 127.

At the C level there is no "unicode bug" and frankly I'm if the view there
is no such thing except in the minds of those who do not understand the
difference between an octet and a codepoint.

In short anyone working on internals or data on the wire which have to
treat series of octets as codepoints of differing representations there is
a need to be able to visually  differentiate strings flagged as utf8 and
those that aren't. Eg, if a c buffer is flagged as utf8 then I need to use
a macro to advance to the next codepoint. If it is not so flagged then I
can increment a pointer. These distinctions matter.

Please do not distract this thread with discussion of "the unicode bug". If
there is one then it is a Perl issue, it is not relevant to debugging the
internals which is what these functions are for.

Thanks,
Yves

>

**Comment by Grinnz** [comment]
This is regarding SV dumps as seen in Devel::Peek, which involve the Perl-logical value of the SV. I am fully aware of the distinction.

**Comment by Tux** [comment]
Do you have example of formatted output before and after?
I am almost sure that my testst in [Data::Peek](https://github.com/Tux/Data-Peek) will start failing, and I will need to add test code for multiple versions. I'm with Yves. Not opposed, but I want to see the impact.

**Comment by sisyphus** [comment]
For me, the major shortcoming of Devel::Peek is its inaccurate presentation of the value in the NV slot.
This is perhaps out of the scope of this thread, but I throw it out there in case it ultimately encourages someone to fix this deficiency.

For anyone unfamiliar with (&& interested in) what I'm alluding to, I present the following example where 2 NVs ($x and $y) contain unequal values, yet Devel::Peek::Dump() presents them as having identical NV values, even though perl itself recognizes the 2 scalars as being unequal:
```
perl -MDevel::Peek -wle '$x = 1.4142135623731; $y = 1.4142135623730951; print "unequal" if $x != $y; Dump $x; Dump $y;'
unequal
SV = NV(0x5642440da7e8) at 0x5642440da800
  REFCNT = 1
  FLAGS = (NOK,pNOK)
  NV = 1.4142135623731
SV = NV(0x5642440da7a0) at 0x5642440da7b8
  REFCNT = 1
  FLAGS = (NOK,pNOK)
  NV = 1.4142135623731
```
(This is on a perl whose nvsize is 8.)
One solution would be to provide that actual NV value in hex ... but there are, of course, other solutions.

This current behaviour doesn't DWIM for me.
I expect accuracy from a function named "Dump".

**Comment by FGasper** [comment]
Before:
```
> perl -MDevel::Peek -e'Dump "hello-é"'
SV = PV(0x7f905800c080) at 0x7f9047812660
  REFCNT = 1
  FLAGS = (POK,IsCOW,READONLY,PROTECT,pPOK)
  PV = 0x7f90575063e0 "hello-\303\251"\0
  CUR = 8
  LEN = 10
  COW_REFCNT = 0

> perl -MDevel::Peek -e'Dump do { my $v = "hello-é"; utf8::upgrade($v); $v }'
SV = PV(0x7fceb300bf50) at 0x7fceb300b448
  REFCNT = 1
  FLAGS = (TEMP,POK,pPOK,UTF8)
  PV = 0x7fceb2406310 "hello-\303\203\302\251"\0 [UTF8 "hello-\x{c3}\x{a9}"]
  CUR = 10
  LEN = 12
```

After:
```
> ./perl -Ilib -MDevel::Peek -e'Dump "hello-é"'
SV = PV(0x7ff28400b890) at 0x7ff285012c58
  REFCNT = 1
  FLAGS = (POK,IsCOW,READONLY,PROTECT,pPOK)
  PV = 0x7ff283f04760 "hello-\xc3\xa9"\0
  CUR = 8
  LEN = 10
  COW_REFCNT = 0

> ./perl -Ilib -MDevel::Peek -e'Dump do { my $v = "hello-é"; utf8::upgrade($v); $v }'
SV = PV(0x7fdd7a80b680) at 0x7fdd7a80ac48
  REFCNT = 1
  FLAGS = (TEMP,POK,pPOK,UTF8)
  PV = 0x7fdd7a605a40 "hello-\xc3\x83\xc2\xa9"\0 [UTF8 "hello-\x{c3}\x{a9}"]
  CUR = 10
  LEN = 12
```

**Comment by FGasper** [comment]
BTW, I’ll happily fix the failing tests once the concerns of @demerphq and others are satisfied.

**Comment by Tux** [comment]
@sisyphus 
```
This is perl 5, version 34, subversion 0 (v5.34.0) built for x86_64-linux-thread-multi-quadmath
```
```
$ perl -MDevel::Peek -wle '$x = 1.4142135623731; $y = 1.4142135623730951; print "unequal" if $x != $y; Dump $x; Dump $y;'
unequal
SV = NV(0x1060960) at 0x1054898
  REFCNT = 1
  FLAGS = (NOK,pNOK)
  NV = 1.4142135623731
SV = NV(0x1060970) at 0x1054850
  REFCNT = 1
  FLAGS = (NOK,pNOK)
  NV = 1.4142135623730951
```
Data::Peek does similar, but is maybe easier to patch without changing CORE behavior
```
perl -MData::Peek -wle '$x = 1.4142135623731; $y = 1.4142135623730951; print "unequal" if $x != $y; DDump $x;DDump $y'
unequal
SV = NV(0x58f290) at 0x516898
  REFCNT = 1
  FLAGS = (NOK,pNOK)
  NV = 1.4142135623731
SV = NV(0x58f2a0) at 0x516850
  REFCNT = 1
  FLAGS = (NOK,pNOK)
  NV = 1.4142135623730951
```

**Comment by Tux** [comment]
@sisyphus -Duselongdouble
```
This is perl 5, version 28, subversion 0 (v5.28.0) built for x86_64-linux-thread-multi-ld
```
```
perl -MData::Peek -wle '$x = 1.4142135623731; $y = 1.4142135623730951; print "unequal" if $x != $y; DDump $x;DDump $y'
unequal
SV = NV(0x1473990) at 0x1438c00
  REFCNT = 1
  FLAGS = (NOK,pNOK)
  NV = 1.4142135623731
SV = PVNV(0x2d1e410) at 0x1438bb8
  REFCNT = 1
  FLAGS = (NOK,pIOK,pNOK)
  IV = 1
  NV = 1.4142135623730951
  PV = 0
```

**Comment by Tux** [comment]
@FGasper Data::Peek uses `pv_display ()` directly, so it will FAIL tests
```
$ perl -MData::Peek -e'DPeek "hello-é"'
PV("hello-\303\251"\0)
$ perl -MData::Peek -e'DPeek do { my $v = "hello-é"; utf8::upgrade($v); $v }'
PV("hello-\303\203\302\251"\0) [UTF8 "hello-\x{c3}\x{a9}"]
```
I think I can live with it. Your change that is. But I have no idea if there are other modules that will fail because of this change. My cpangrep-foo is bad.

**Comment by sisyphus** [comment]
@Tux, the same problem exists with -Duselongdouble and -Dusequadmath builds of perl, though it requires different values to expose it.

On perls whose NV is the 80-bit extended precision long double, there are (UIM) 92 different NV values that will be presented by Devel::Peek::Dump() and the print() function as <code>1.41421356237309505</code>.
These are the values <code>0xb.504f333f9de6462p-3 ..  0xb.504f333f9de64bdp-3</code> where the incrementation is done in steps of 1 ULP.
One of those 92 is the NV returned by <code>sqrt(2.0)</code>.

On perls whose NV is the full 128-bit quad type (either "long double" or "__float128"), there are (UIM) 52 different NV values that will be presented by Devel::Peek::Dump() and the print() function as <code>1.4142135623730950488016887242097</code>.
These are the values <code>0x1.6a09e667f3bcc908b2fb1366ea86p+0 ..  0x1.6a09e667f3bcc908b2fb1366eab9p+0</code>,
again incrementing by 1 ULP.
One of those 52 is the NV returned by <code>sqrt(2.0)</code>.

Anyway ... I think we're all aware of perl's limitations regarding interpolation of NVs.
I just wanted to remind people that the same limitations currently apply also to Devel::Peek::Dump's display of the value in the NV slot. 
I actually got suckered by this a few weeks ago. when I used Devel::Peek::Dump to verify that 2 scalars were, in fact, both NVs.
They were, and I also noticed that their NV slot values were reported as identical.
After another 20 minutes of head scratching, I eventually realized that the 2 NV slots actually contained different values.
I should have known better, but I had allowed myself to be fooled by the power of suggestion at a time when I wasn't "properly on my guard".

**Comment by haarg** [comment]
I've also been annoyed by the inaccuracies when dumping float values. But that seems like it should go in a separate issue.

**Comment by FGasper** [comment]
@Tux I’ve pushed up a change that scopes the change specifically to sv_dump(); the public pv_display() should now work as it did. Thoughts?

**Comment by khwilliamson** [comment]
On 1/29/22 04:49, sisyphus wrote:
> For me, the major shortcoming of Devel::Peek is its inaccurate 
> presentation of the value in the NV slot.
> This is perhaps out of the scope of this thread, but I throw it out 
> there in case it ultimately encourages someone to fix this deficiency.

For any chance of this happening, please submit a ticket

**Comment by FGasper** [comment]
@demerphq Have you been able to look more at this?

For reference, the difference this makes is:
```
> ./perl -Ilib -MDevel::Peek -e'Dump "\x80\x81"'
SV = PV(0x7fb57480b890) at 0x7fb57580e658
  REFCNT = 1
  FLAGS = (POK,IsCOW,READONLY,PROTECT,pPOK)
  PV = 0x7fb574505460 "\x80\x81"\0
  CUR = 2
  LEN = 10
  COW_REFCNT = 0
```
(Note the absence of octal.)

… versus this in upstream:
```
> perl -MDevel::Peek -e'Dump "\x80\x81"'
SV = PV(0x7fa56380c680) at 0x7fa56400ce60
  REFCNT = 1
  FLAGS = (POK,IsCOW,READONLY,PROTECT,pPOK)
  PV = 0x7fa563705460 "\200\201"\0
  CUR = 2
  LEN = 10
  COW_REFCNT = 0
```
UTF8-flagged strings look thus:
```
> ./perl -Ilib -MDevel::Peek -e'Dump do { utf8::upgrade(my $v = "\x80\x81"); $v }'
SV = PV(0x7fdf1a80b680) at 0x7fdf1a80ac48
  REFCNT = 1
  FLAGS = (TEMP,POK,pPOK,UTF8)
  PV = 0x7fdf1a505800 "\xc2\x80\xc2\x81"\0 [UTF8 "\x{80}\x{81}"]
  CUR = 4
  LEN = 10
```
I, at least, find the hex much easier to read.

**Comment by demerphq** [comment]
On Fri, 11 Feb 2022, 21:41 Felipe Gasper, ***@***.***> wrote:

> @demerphq <https://github.com/demerphq> Have you been able to look more
> at this?
>
My apologies. As you know from private mails that I had some issues with
testing your patch and couldn't get it to produce the output you expected,
for some reason git didnt checkout your branch, but rather created a new
branch pointing at HEAD at the time, so I was testing the wrong thing[1]
And then I got covid[2], and it completely slipped my mind until your mail
yesterday. Sorry about that.

I have checked that it does not make using the regex debugger harder and I
am relatively satisfied on that point I think.

But I see an inconsistency in the result. Null bytes are still being output
as octal:

$ ./perl -Ilib -MDevel::Peek -e'my $s="\x{0}0\x{0}a\x{DF}a\x{DF0}b";
Dump($s); utf8::upgrade($s); Dump($s)'
SV = PV(0x561268ad3eb8) at 0x561268b04968
  REFCNT = 1
  FLAGS = (POK,IsCOW,pPOK,UTF8)
  PV = 0x561268b10668 "\0000\0a\xc3\x9fa\xe0\xb7\xb0b"\0 [UTF8
"\x{0}0\x{0}a\x{df}a\x{df0}b"]
  CUR = 11
  LEN = 13
  COW_REFCNT = 1
SV = PV(0x561268ad3eb8) at 0x561268b04968
  REFCNT = 1
  FLAGS = (POK,IsCOW,pPOK,UTF8)
  PV = 0x561268b10668 "\0000\0a\xc3\x9fa\xe0\xb7\xb0b"\0 [UTF8
"\x{0}0\x{0}a\x{df}a\x{df0}b"]
  CUR = 11
  LEN = 13
  COW_REFCNT = 1

I also personally find it difficult to read, especially as it is using
lower case unbraced hex for the octet form.

I would feel a lot better if it used upper case hex digits. So that "\xc3a"
would instead be output as "\xC3a".  I guess it's a taste thing, but I find
it difficult to read "\xef" compared to "\xEF", to me it would make it more
obvious what was going on. I know the unicode output uses lower case, but
it also has braces so the readability isnt a problem.

Anyway, I strongly think that NULL should be consistent with other
codepoints.

But my concerns about the impact on debugging the regex engine are
withdrawn, if you fixed the null as octal I would be ok with it, even if it
does make a lot of strings much longer, which I think was the reason it
used octal in the first place (shortest unambiguious way to represent
binary). I definitely agree its easier to decode hex than octal. Once again
sorry for the delay.

cheers,
Yves
[1] Maybe in future you could create your branches with a two-tuple, eg
"felipe/sv_dump_to_hex". I think that might have helped me realize my
mistake.  Pretty well all the branches in the main repo are two-tuple form.
[2] Omnicron, mild, tested negative this morning.

**Comment by FGasper** [comment]
@demerphq
- I’ve made NULs consistent with the rest. (I do wonder why `PERL_PV_ESCAPE_DWIM` makes the NUL octal, but hey.)
- I’ve switched to upper-case hex.

**Comment by khwilliamson** [comment]
Who still has a problem with merging this?

**Comment by demerphq** [comment]
On Fri, 18 Feb 2022, 08:52 Karl Williamson, ***@***.***>
wrote:

> Who still has a problem with merging this?
>

Just be clear Felipe resolved all my concerns and I am fine to merge it.

Yves

>

**Comment by khwilliamson** [comment]
No one said they had a problem with merging this, so now doing that

**Comment by Tux** [comment]
@atoomic sorry for the late reply, but I'm oké with this. I just tested with blead which has this PR merged and all is fine.
